subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/DefaultFileSystemSnapshotter.java

```diff
@@ -276,4 +273,24 @@ public void visitDirectoryTree(DirectoryFileTree directoryTree) {
             return roots;
         }
     }
+
+    private class InternedPathHolder {
+        private final String nonInternedPath;
+        private String internedPath;
+
+        public InternedPathHolder(String nonInternedPath) {
+            this.nonInternedPath = nonInternedPath;
+        }
+
+        public String getInternedPath() {
+            if (internedPath == null)  {
+                internedPath = stringInterner.intern(nonInternedPath);
+            }
+            return internedPath;
+        }
+
+        public String getNonInternedPath() {
+            return nonInternedPath;
+        }
+    }
```

Comment:
Would it make sense to store a single version of the string and remember if it's already interned instead? I'm guessing that would take less memory. Not sure if it's worth it though.

Also, I think `asInterned()` vs `asNonInterned()` would capture what we are trying to do here. So would calling the type `InternableString`. WDYT?

Reply:
I renamed the thing to `InternableString` with `asInterned` and `asNonInterned` methods.